crates/environ/src/module.rs

```diff
@@ -473,6 +565,65 @@ pub struct TableInitializer {
     pub elements: Box<[FuncIndex]>,
 }
 
+/// Table initialization data for one table.
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub enum TableInitialization {
+    /// "Segment" mode: table initializer segments, possibly with
+    /// dynamic bases, possibly applying to an imported memory.
+    ///
+    /// Every kind of table initialization is supported by the
+    /// Segments mode.
+    Segments {
+        /// The segment initializers. All apply to the table for which
+        /// this TableInitialization is specified.
+        segments: Vec<TableInitializer>,
+    },
+
+    /// "FuncTable" mode: a single array, with a function index or
+    /// null per slot. This is only possible to provide for tables
+    /// that a module itself defines, and only when the initializer
+    /// segments have statically-knowable bases (i.e., not dependent
+    /// on global values).
+    ///
+    /// This mode facilitates lazy initialization of the table. It is
+    /// thus "nice to have", but not necessary for correctness.
+    FuncTable {
+        /// An array of function indices or None (meaning no initialized
+        /// value, hence null by default). Array elements correspond
+        /// one-to-one to table elements; i.e., `elements[i]` is the
+        /// initial value for `table[i]`.
+        elements: Vec<Option<FuncIndex>>,
+    },
+}
+
+impl TableInitialization {
+    /// If in Segments mode, get a slice of the segment initializers,
+    /// otherwise None.
+    pub fn as_segments(&self) -> Option<&[TableInitializer]> {
+        match self {
+            Self::Segments { segments } => Some(&segments[..]),
+            _ => None,
+        }
+    }
+
+    /// If in Segments mode, get a mutable borrow of the vector of
+    /// segment initializers, otherwise None.
+    pub fn as_segments_mut(&mut self) -> Option<&mut Vec<TableInitializer>> {
+        match self {
+            Self::Segments { segments } => Some(segments),
+            _ => None,
+        }
+    }
+
+    /// If in FuncTable mode, get the array of Option<FuncIndex> elements.
+    pub fn as_func_table(&self) -> Option<&[Option<FuncIndex>]> {
+        match self {
+            Self::FuncTable { elements } => Some(&elements[..]),
+            _ => None,
+        }
+    }
+}
```

Comment:
Personally I try to avoid methods like this for not-so-commonly-used types since specifically asking for one variant almost defeats the purpose of exhaustive matching in Rust. In other words we'd lose any benefit where if a variant is added to this `enum` we can't rely on the compiler to tell us to go update all the various places, instead we have to audit all callers of these methods to determine if they need to handle the new case.

If possible can these helpers get removed in favor of a `match` on call-sites with comments for why one or the other case is ignored?

Reply:
Done!